File: BFS/Leetcode133.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
    if(!node) {
            return node;
        }
    const que = [];
    const map = new Map();
    // HashSet<Node> visited = new HashSet<Node>
    que.push(node);
    map.set(node, new Node(node.val));

    for(let i = 0; i < que.length; i++) {
        const temp = que[i];
        temp.neighbors.forEach((neighbor) => {
            if(!map.has(neighbor)) {
                map.set(neighbor, new Node(neighbor.val));
                que.push(neighbor);
            }
        })
    }

    // connect all nodes
    for(let i = 0; i < que.length; i++) {
        const origin = que[i];
        const copy = map.get(origin);
        origin.neighbors.forEach((neighbor) => {
            copy.neighbors.push(map.get(neighbor));
        });
    }

    return map.get(node);
};

File: BFS/Leetcode133.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) {
            return node;
        }
        List<Node> que = new ArrayList<Node>();
        HashMap<Node, Node> map = new HashMap<Node, Node>();
        // HashSet<Node> visited = new HashSet<Node>
        que.add(node);
        map.put(node, new Node(node.val));
        
        for(int i = 0; i < que.size(); i++) {
            Node temp = que.get(i);
            for(Node neighbor: temp.neighbors) {
                if(!map.containsKey(neighbor)) {
                    map.put(neighbor, new Node(neighbor.val));
                    que.add(neighbor);
                }
            }
        }
        
        // connect all nodes
        for(int i = 0; i < que.size(); i++) {
            Node origin = que.get(i);
            Node copy = map.get(origin);
            for(Node neighbor: origin.neighbors) {
                copy.neighbors.add(map.get(neighbor));
            }
        }
        
        return map.get(node);
    }
}